src/librbd/mirror/snapshot/RemoveGroupSnapshotRequest.cc

```diff
@@ -0,0 +1,167 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "include/ceph_assert.h"
+#include "common/Cond.h"
+#include "common/dout.h"
+#include "common/errno.h"
+#include "common/ceph_context.h"
+#include "cls/rbd/cls_rbd_client.h"
+#include "librbd/Operations.h"
+#include "librbd/Utils.h"
+#include "librbd/api/Utils.h"
+#include "librbd/mirror/snapshot/RemoveGroupSnapshotRequest.h"
+
+#define dout_subsys ceph_subsys_rbd
+#undef dout_prefix
+#define dout_prefix *_dout << "librbd::mirror::snapshot::RemoveGroupSnapshotRequest: " << this \
+                           << " " << __func__ << ": "
+
+namespace librbd {
+namespace mirror {
+namespace snapshot {
+
+using util::create_rados_callback;
+using util::create_context_callback;
+
+template <typename I>
+void RemoveGroupSnapshotRequest<I>::send() {
+  ldout(m_cct, 10) << dendl;
+
+  remove_group_snapshot(*m_group_snap);
+}
+
+template <typename I>
+void RemoveGroupSnapshotRequest<I>::remove_group_snapshot(
+                              cls::rbd::GroupSnapshot group_snap) {
+  ldout(m_cct, 10) << "group snap id: " << group_snap.id << dendl;
+
+  //TODO: Handle dynamic group membership
+  if (!m_image_ctxs->empty() && m_image_ctx_map.empty()) {
+    for (size_t i = 0; i < m_image_ctxs->size(); ++i) {
+      ImageCtx *ictx = (*m_image_ctxs)[i];
+      m_image_ctx_map[ictx->id] = ictx;
+    }
+  }
+
+  auto ctx = create_context_callback<
+      RemoveGroupSnapshotRequest,
+      &RemoveGroupSnapshotRequest<I>::handle_remove_group_snapshot>(this);
+
+  m_group_snap_id = group_snap.id;
+
+  C_Gather *gather_ctx = new C_Gather(m_cct, ctx);
+  for (auto &snap : group_snap.snaps) {
+    if (snap.snap_id == CEPH_NOSNAP) {
+      continue;
+    }
+    ImageCtx *ictx = nullptr;
+    if (m_image_ctx_map.find(snap.image_id) != m_image_ctx_map.end()) {
+      ictx = m_image_ctx_map[snap.image_id];
+    }
+
+    if (!ictx) {
```

Comment:
doesn't this indicate that the group image snapshot doesn't exist?

Reply:
Yes, which is why it log and continue to the next image's snapshot.

Reply:
In that case logging it as "group image snapshot already removed" is more insightful IMO